thrust/detail/get_iterator_value.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+/*
+ *  Copyright 2008-2016 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <thrust/detail/config.h>
+
+namespace thrust {
+namespace detail {
+
+// get_iterator_value specialization on iterators
+// --------------------------------------------------
+// it is okay to dereference iterator in usual way
+template<typename DerivedPolicy, typename Iterator>
+__host__ __device__
+typename thrust::iterator_traits<Iterator>::value_type
+get_iterator_value(thrust::execution_policy<DerivedPolicy> &, Iterator it)
```

Comment:
I understand. However, this fix doesn't change the previously existed behaviour, but fixes a bug when a raw pointer is passed to max_element.

That being said, replacing `*it` with 

      typename thrust::iterator_traits<Iterator>::value_type value;
      thrust::copy(exec, it,it+1, &value); 
      return value; 

makes the reproducer to die with

    terminate called after throwing an instance of 'thrust::system::system_error'
      what():  an illegal memory access was encountered
    Aborted (core dumped)

when used with transform_iterator and I wasn't able to quickly nail  down the issue. I agree the get_value_iterator needs to be enhanced to respect customer user policy, but it may take more time.

I suggest either accepting this fix or reverting https://github.com/thrust/thrust/pull/777 to make sure the user codes are not broken, and then filing a separate bug that requires get_iterator_value to respect user policy.